packages/flutter_tools/lib/src/test/runner.dart

```diff
@@ -89,14 +124,59 @@ Future<int> runTests(
 
   // Call package:test's main method in the appropriate directory.
   final Directory saved = fs.currentDirectory;
+
+  void printReloadMessage() {
+    printStatus(
+      "Press 'r' to rerun your tests, 'q' to quit",
+      color: TerminalColor.red
+    );
+  }
+
   try {
     if (workDir != null) {
       printTrace('switching to directory $workDir to run tests');
       fs.currentDirectory = workDir;
     }
 
     printTrace('running test package with arguments: $testArgs');
-    await test.main(testArgs);
+    await test.runTests(testArgs);
+
+    if (watchTests) {
+      final DirectoryWatcher directoryWatcher = DirectoryWatcher(saved.path);
+
+      directoryWatcher.events.listen((WatchEvent event) {
+        if (event.path.endsWith('.dart')) {
+          invalidatedFiles.add(event.path);
+        }
+      });
+
+      printReloadMessage();
+
+      terminal.singleCharMode = true;
+      await for (String char in terminal.keystrokes.asBroadcastStream()){
+        if (char == 'r') {
+          final Status status = logger.startProgress('Recompiling test files...\n', timeout: null);
+          await compileTestFiles();
```

Comment:
Do we need to worry about any of these methods throwing?

Reply:
Well, I don't think so, at least not in a way that would not be expected behavior.  Actually one exception, listening to the terminal.keystrokes could throw with a StateError since there is the ever present threat of a race condition with some dependency to listen to the stdin stream(which is singlecast).

Reply:
A good thing to note, is that this section is now covered in a try/catch that covers the StateError.  So I think we should be good here.